Reduce allocations caused by logging.

MarchingCube opened the pull request:
## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Back with more memory pressure fixes.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
Logging functions allocate temporary arrays, box values and generally do busywork even with logging disabled.

![image](https://user-images.githubusercontent.com/2588062/66000566-fdc52700-e49f-11e9-96aa-518f51e0e83d.png)

![image](https://user-images.githubusercontent.com/2588062/66000632-22210380-e4a0-11e9-9edc-f4e0d75a3446.png)


## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
Significant decrease in amount of `object[]` and boxed value types being allocated.

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->
- implement 4 new logging functions that take from 0 to 3 arguments using generics. To avoid extra work I only implemented base `Log` function so you have to pass verbosity (maybe we need some T4 template for that or something, but I really didn't want to write this by hand). Only 3 arguments since Serilog does not support more anyway and you end up with params.
- allow for checking if log level is enabled at all to avoid doing avoidable work (like measuring layout time for no reason).

After suggestion from @kekekeks I also added two new conditional logging functions:
- `TryGetLogger` that returns nullable parametrized logger
- `TryGetLogger` that returns boolean and has output logger parameter.

Both can be used when actual logging is expensive (multiple parameters, something has to be calculated) as we don't want to bog down the performance due to the logging.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

Part of https://github.com/AvaloniaUI/Avalonia/issues/2919

MarchingCube committed:
Reduce allocations caused by logging.

kekekeks commented on src/Avalonia.Base/AvaloniaObject.cs:
```diff
@@ -326,21 +326,25 @@ public bool IsSet(AvaloniaProperty property)
-                    "Bound {Property} to {Binding} with priority LocalValue", 
-                    property, 
-                    description);
+                if (Logger.IsEnabled(LogEventLevel.Verbose))
```
We should check for both log level and log area, I think.

kekekeks replied:
I think the better API would be a `TryGetLogger` method that returns `null` when logging is not configured, so the logging call would look like

```cs
Logger.TryGetLogger(LogEventLevel.Verbose, LogArea.Property)
    ?.Log(this, "Bound {Property} to {Binding} with priority LocalValue", property,  GetDescription(source));

```
which is more clear than `if` block and passing the log level twice.

MarchingCube replied:
There is no way of checking if area is enabled it only decorates log output.
And I agree `TryGetLogger` API is nicer, will change mine to use something like that.

MarchingCube replied:
Added new API as well.

MarchingCube committed:
Cleanup conditional logging API.

kekekeks commented on src/Avalonia.Base/AvaloniaObject.cs:
```diff
@@ -406,7 +414,7 @@ internal void PriorityValueChanged(AvaloniaProperty property, int priority, obje
                 RaisePropertyChanged(property, oldValue, newValue, (BindingPriority)priority);
 
-                Logger.Verbose(
+                Logger.TryGetLogger(LogEventLevel.Verbose)?.Log(
```
This syntax would look way better in other changed parts of the code. We should probably also remove the direct way of writing to log, so people won't accidentally cause unnecessary allocations.

MarchingCube replied:
Problem is mostly caused by params array allocation.
I can use this pattern in one other place, but for a few other places I have to use boolean returning variant since we also do extra work if the logger is enabled (like fetching the description).


I think I might just add more generic parameters and expand to params in the logger itself. Maybe I should remove params overload completely? Otherwise somebody adding one extra parameter to the existing logging call can unknowingly introduce allocations.

kekekeks replied:
:+1: for removing `params`

MarchingCube committed:
Cleanup.

MarchingCube committed:
Remove direct logging functions.

MarchingCube commented:
@kekekeks Removed direct logging access and all convenience functions.
I also changed name from `Logger.TryGetLogger` to `Logger.TryGet` as I think it is obvious that we are getting a logger.

Removed `params` from the main API, if somebody wants to log `params` variant they can use public `ILogSink` that is still exposed from the `Logger` class.

To keep basic compatibility I extended `ParametrizedLogger`  to take up to 6 generic arguments, everything past 3rd one goes to `ILogSink` `params` API. But thanks to conditional logging if people have no logger enabled it won't allocate anything.

kekekeks commented on src/Avalonia.Base/AvaloniaObject.cs:
```diff
@@ -326,21 +326,24 @@ public bool IsSet(AvaloniaProperty property)
-                    "Bound {Property} to {Binding} with priority LocalValue", 
-                    property, 
-                    description);
+                if (Logger.TryGet(LogEventLevel.Verbose, out var logger))
```
```cs
Logger.TryGet(LogEventLevel.Verbose, out var logger)?.Log(
                        LogArea.Property,
                        this,
                        "Bound {Property} to {Binding} with priority LocalValue",
                        property,
                        GetDescription(source);
```

kekekeks reviewed (approved):
LGTM

grokys committed:
Merge branch 'master' into alloc-logging

MarchingCube committed:
Final review fixes.
